packages/components/src/components/amount-input/amount-input.tsx

```diff
@@ -0,0 +1,137 @@
+        setSelection({ selectionStart: updated_caret_position, selectionEnd: updated_caret_position });
+    }, [value]);
+
+    const onChangeHandler: React.ComponentProps<typeof Input>['onChange'] = e => {
+        if (!target) setTarget(e.target);
+        let newValue = value;
+        if (!is_pasting) {
+            // handle ATM typing:
+            // remove all characters that are not digit / point / comma:
+            const input_value = e.target.value.replace(/[^\d.,]/g, '');
+            if (input_value.replace(/[.,]/g, '').replace(/^0+/g, '').length <= max_digits)
+                newValue = Number(input_value.replace(/[.,]/g, '')) / Math.pow(10, decimal_places);
+        } else {
+            // handle pasting:
+            const selection_length = selection.selectionEnd - selection.selectionStart;
+            const pasted_string_length = e.target.value.length + selection_length - displayNumber(value).length;
+            const pasted_string = e.target.value.substring(
+                selection.selectionStart,
+                selection.selectionStart + pasted_string_length
+            );
+            // remove all characters that are not digit / point / comma:
+            const input_value = e.target.value.replace(/[^\d.,]/g, '');
+            // understand the value user wants to paste:
+            const pasted_value = pasted_string
+                .replace(/[^\d.,]/g, '') // remove all characters that are not digit / point / comma
+                .replace(/[,.](?=.*[,.])/g, '') // leave only the last point / comma
+                .replace(',', '.'); // make the last point / comma separator a point
+            if (pasted_value.replace('.', '')) {
+                // if the value is a valid non-empty string, handle the two scenarios:
+                if ((value === 0 && caret_right_offset === 0) || selection_length === displayNumber(value).length) {
+                    // handle pasting when there's nothing entered before it, or it is overridden (intention: reset value):
+                    newValue = Number(
+                        pasted_value.substring(
+                            0,
+                            pasted_value.includes('.') ? max_digits + 1 : max_digits - decimal_places
+                        )
+                    );
+                } else if (input_value.replace(/[.,]/g, '').replace(/^0+/g, '').length <= max_digits) {
+                    // handle pasting when there's something entered before it and there's space for the pasted value (intention: add to value):
+                    newValue = Number(input_value.replace(/[.,]/g, '')) / Math.pow(10, decimal_places);
+                }
+            }
+        }
+        setValue(newValue);
+        setIsPasting(false);
+        onChange?.(newValue);
+    };
+
+    const inputActionHandler = (e: {
+        currentTarget: { value: string; selectionStart: number | null; selectionEnd: number | null };
```

Comment:
Don't know how to combine `React.ComponentProps<typeof Input>['onMouseDown']`, `React.ComponentProps<typeof Input>['onMouseUp']`, and `React.ComponentProps<typeof Input>['onKeyDown']` here 🙁

Reply:
Have asked ChatGPT to help with this, and it produced a better typing.
I tried using the same type previously, but I was typing it as `React.ComponentProps<typeof Input>['onMouseDown' | 'onMouseUp' | 'onKeyDown']`, and this was producing an error. ChatGPT told me that I can't use `|` in square brackets like this and suggested the current typing.

Reply:
<img width="811" alt="Screenshot 2023-05-19 at 12 56 46" src="https://github.com/binary-com/deriv-app/assets/119863957/06e13d4f-a258-4419-b584-05a0c95bc76f">